pkg/spdk/engine.go

```diff
@@ -1229,6 +1346,415 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
+	requireExpansion, err := e.startExpansion(size, replicaClients)
+	if err != nil {
+		return errors.Wrap(err, "startExpansion failed")
+	}
+	if !requireExpansion {
+		e.log.Info("No need to expand engine")
+		e.SpecSize = size
+		return nil
+	}
+
+	var (
+		expanded  = false
+		engineErr error
+	)
+
+	// engineErr will be set when the engine failed to do any non-recoverable operations.
+	defer func() {
+		if retErr != nil {
+			e.log.WithError(engineErr).Errorf("Engine %s failed to expand", e.Name)
+			e.lastExpansionError = errors.Wrap(retErr, "engine failed to expand expansion").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		if engineErr != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+			}
+			e.ErrorMsg = engineErr.Error()
+			e.log.WithError(engineErr).Errorf("Engine %s under non-recoverable operation during expansion", e.Name)
+
+			if errUpdateLogger := e.log.UpdateLogger(logrus.Fields{
+				"replicaStatusMap": e.ReplicaStatusMap,
+			}); errUpdateLogger != nil {
+				e.log.WithError(errUpdateLogger).Warn("Failed to update logger with replica status map during engine creation")
+			}
+
+			e.lastExpansionError = errors.Wrap(retErr, "engine under non-recoverable operation").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		e.finishExpansion(expanded, size)
+	}()
+
+	// prepareRaidForExpansion checks if RAID exists, suspends frontend if needed, and deletes the RAID bdev.
+	isSuspended, bdevRaidUUID, err := e.prepareRaidForExpansion(spdkClient)
+	if isSuspended {
+		defer func() {
+			if frontendErr := e.initiator.Resume(); frontendErr != nil {
+				frontendErr = errors.Wrapf(frontendErr, "failed to resume NVMe initiator during engine %s", e.Name)
+				if engineErr != nil {
+					engineErr = errors.Wrapf(engineErr, "original error; resume failed: %v", frontendErr)
+				} else {
+					engineErr = frontendErr
+				}
+			}
+		}()
+	}
+	if err != nil {
+		engineErr = errors.Wrap(err, "prepare raid for expansion failed")
+		retErr = engineErr
+		return retErr
+	}
+
+	// expand replicas
+	if err := e.expandReplicas(replicaClients, spdkClient, size); err != nil {
+		engineErr = err
+		retErr = err
+		return retErr
+	}
+
+	// recreate RAID and reconnect frontend
+	if err := e.reconnectFrontend(spdkClient, bdevRaidUUID); err != nil {
+		engineErr = errors.Wrap(err, "reconnectFrontend failed")
+		retErr = engineErr
+		return retErr
+	}
+	e.log.Info("Expanding engine complete")
+
+	expanded = true // which could be true even in partial success
+	return nil
+}
+
+func (e *Engine) startExpansion(size uint64, replicaClients map[string]*client.SPDKClient) (requireExpansion bool, err error) {
+	if e.isExpanding {
+		return false, fmt.Errorf("expansion is in progress")
+	}
+
+	if e.IsRestoring {
+		return false, fmt.Errorf("restoring is in progress")
+	}
+
+	if e.SpecSize > size {
```

Comment:
Change to below to avoid unnecessary `else` for simplifying code readability.

```go
	if e.SpecSize > size {
		return false, fmt.Errorf("cannot expand engine to a smaller size %v, current size %v", size, e.SpecSize)
	}
    if e.SpecSize == size {
		e.log.Infof("Engine already at requested size %v, skipping expansion", size)
		return false, nil // no need to expand
	}
```